### ISSUE TITLE
Fail to run with iupac codes

### ISSUE THREAD

**Issue by erick-dorlass** [PoonLab/sierra-local#72]
Hello,
I'm using sierra-local for predicting HIV-1 resistance in sequences generated by a low frequency calls pipeline. This made necessary to use IUPAC codes in some mutation sites. Using some sequences, even with iupac codes, the program runs fine. However, when using sequences that contain a larger number of IUPAC codes nucleotides, I receive this error:

| 2023-01-30T16:30:38.339-03:00 | searching path /usr/local/lib/python3.6/dist-packages/sierralocal/data/HIVDB*.xml
| 2023-01-30T16:30:38.339-03:00 | searching path /usr/local/lib/python3.6/dist-packages/sierralocal/data/apobec*.tsv
| 2023-01-30T16:30:38.339-03:00 | HIVdb version 8.8
| 2023-01-30T16:30:38.339-03:00 | Found NucAmino binary /usr/local/lib/python3.6/dist-packages/sierralocal/bin/nucamino-linux-amd64
| 2023-01-30T16:30:38.339-03:00 | Aligned /tmp/scratch/path_to_file/HIVRes/dd186464-42ee-4232-b3d0-6d4e1ddcb863/call-GenerateConsensusSensitive015/sample.fasta
| 2023-01-30T16:30:38.342-03:00 | Traceback (most recent call last):
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/bin/sierralocal", line 11, in <module>
| 2023-01-30T16:30:38.342-03:00 | exit_code = main.main()
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/main.py", line 179, in main
| 2023-01-30T16:30:38.342-03:00 | cleanup=args.cleanup, forceupdate=args.forceupdate)
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/main.py", line 122, in sierralocal
| 2023-01-30T16:30:38.342-03:00 | file_trims, subtypes = scorefile(input_file, algorithm)
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/main.py", line 51, in scorefile
| 2023-01-30T16:30:38.342-03:00 | aligner.get_mutations(result)
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/nucaminohook.py", line 283, in get_mutations
| 2023-01-30T16:30:38.342-03:00 | subtype = self.typer.getClosestSubtype(record['Sequence'], offset)
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/subtyper.py", line 102, in getClosestSubtype
| 2023-01-30T16:30:38.342-03:00 | dists = self.getDistances(sequence, offset)
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/subtyper.py", line 75, in getDistances
| 2023-01-30T16:30:38.342-03:00 | dists[header] = self.uncorrectedDistance(sequence, reference[offset:])
| 2023-01-30T16:30:38.342-03:00 | File "/usr/local/lib/python3.6/dist-packages/sierralocal/subtyper.py", line 58, in uncorrectedDistance
| 2023-01-30T16:30:38.342-03:00 | if not ref[idx] in self.ambiguities[nuc]:
| 2023-01-30T16:30:38.342-03:00 | KeyError: ' '

The same sequence runs fine in the sierra web service. Is there something I could do?

**Comment by WilliamZekaiWang** [comment]
Hi, I couldn't replicate your issue

I've tried running an RT sequence with ambiguous nucleotides on `HIVdb version 8.8` and `HIVdb version 9.4` and didn't get any error messages.

could you try using the `HIVdb version 9.4` through the submodule to see if this issue still occurs?

**Comment by ArtPoon** [comment]
No follow up from user, closing

**Comment by WilliamZekaiWang** [comment]
Hi Erick,

No worries about the time. 

I ran this sequence through the new alignment program we are swapping to--listed in #71--and it successfully aligned your sequence.

This issue should get fixed in the incoming PR

**Comment by erick-dorlass** [comment]
Thank you for such a quick answer.
I'll wait for the incoming PR.

**Comment by erick-dorlass** [comment]
Hello,

I have tried with the current version in main containing the new alignment program, and indeed I don't get the error anymore. However, the total score of some ARV are wrong, with mutations that are not there due to lack of coverage.
For example, the mutation N348I in RT don't exist, the sequence has an N in this position, wich is confirmed in the list of mutations in RT, but the mutation is still called for the score of NVP drug. Many more of this can be found along the genome, specially in the IN gene.

**Comment by WilliamZekaiWang** [comment]
I will look further into this. 

While testing this program I didn't get to test it much on IN, which could explain the many errors.

**Comment by WilliamZekaiWang** [comment]
There are incorrectly labelled APOBEC mutations which I have found and working on a fix for. This is likely due to an outdated version of the `apobec_drm.json` file. 

Regarding incorrectly identifying mutations, I compared sierralocal, running on postalign, with sierrapy, and the output position and AA matched. Could you elaborate on where you are comparing the outputs so I can have a deeper look?

**Comment by erick-dorlass** [comment]
Hello,

Thank you for the investigation. I'll wait for the fix on the `apobec_drm.json` file.

To assure that we are using the same sequence for these tests, I'll repost it here:

```
>consensus
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNCCAGAGCCAACAGCC
CCACCAGCAGAGAGCTTCAGGTTCGGAGAGGAGATAACCCCCTCTCMGAAGCAGGAACAG
AAGGAAGAGGGAAYATACCCTCCTTTAGCTTCCCTCAAATCACTCTTTGGCAACGACCMG
TARTCACARTAAAAGTAGRGGGACAGCTAAGGGAAGCTTTATTAGAYACAGGAGCAGATG
ATACAGTATTMGAAGACRTAGATTTACCAGGAAAMTGGAAGCCAAAAATAATAGGGGGAR
TTGGAGGTTTTSTCARAGTAAARCAGTATGATAACRTACWCATAGAAATTTGTGGRCAYA
ARGTTACAGGTACAGTAGTAGTAGGACCTACACCTRYSAAYATAATTGGAAGAAATCTGT
TGACTCAGCTGGGTTGCACTTTAAATTTTCCCATWAGTCCTATTGAAACTGTACCAGTAA
AATTAAAGCCAGGAATGGATGGCCCAAAAGTTAAACAATGGCCATTGACAGAAGAAAAAA
TAAAAGCATTAACAGAAATATGTACAGAAATGGAAAAAGAAGGGAAAATTTCAAAAATTG
GGCCTGAAAATCCATAYAATACTCCARTATTTGCCATAAAGAAAAARRACAGTWCTAGAT
GGAGAAAATTAGTAGATTTCAGRGAACTTAATAAAAGAACTCAAGATTTTTGGGAGGTTC
AATTAGGAATACCGCATCCTGCGGGATTAMAAAAGAAAAAATCAGTAACAGTACTGGATG
TGGGTGATGCATATTTTTCAGTTCCCTTAGATGAAGAYTTCAGGAAGTATACTGCATTTA
CCATACCTAGTACMAATAATGAGACACCRGGGATTAGATATCAGTACAATGTGCTTCCAC
AAGGATGGAAAGGATCACCAGCAATATTCCAAAGCAGCATGACAAAAATCTTAGAGCCCT
TTAGAAAACAAAACCCGGACATAGTGATYTATCAATAYGTGGATGATTTGTATGTAGGAT
CTGATYTAGAAATAGGGCAGCATAGAACAAAAWWRAGGAACTGAGACARCATCTRTTGAC
GTGGGGATTGACCACACCAGATMAAAAACACCAGAAAGAACMTCCATTTCTTTGGATGGG
KTATGAACTCCATCCTGATAAATGGACAGTACAGCCTATAGAGCTGCCAGAAAAGGAYAG
CTGGACTGTCAATGACATACAGAAGTTAGTGGGAAAATTRAATTGGGCAAGTCAGATTTA
CCCAGGGATTAARGTAAAGCAATTATGTAGACTCCTTARGGGAACCAAGGCRCTAACAGA
AGTAGTACCMCTAACAAAAGARGCAGAGTTAGAACTGGCAGAAAACAGGGAAATTCTAAA
GGAACCAGTACATGGAGTGTATTATGACCCATCAAAAGACNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTTCGGGATY
AGAAGTAAAYATAGTRACAGACTCACAGTATGCATTAGGAATCATTCAAGCACAACCAGA
TAAGAGTGAATCAGAGATAGTCAATCAAATAATAGAGCAATTAATAACAAAGGAAAAGGT
CTACCTGTCATGGGTACCAGCACACAAAGGAATTGGAGGAAATGAACAAGTAGATAAATT
AGTCAGTACTGGAATCAGAAAAGTACTATTTTTAGATGGAATAGATAAAGCCCAAGAAGA
ACATGAGAAGTATCACAGTAATTGGAGGGCAATGGCCAGTGATTTTAACCTGCCACCTGT
GGTAGCAAAAGAAATAGTAGCCAGCTGTGATAAATGTCAGCTAAAAGGAGAAGCCATGCA
TGGACAAGTAGACTGTAGTCCAGGAATATGGCAACTAGATTGTACACATTTAGAAGGAAA
AGTTATCCTGGTAGCAGTCCATGTAGCTAGTGGATACCTAGAAGCAGAAGTTATCCCAGC
AGAAACAGGACAAGAAACAGCCTACTTCATACTAAAGTTAGCAGGAAGATGGCCAGCAAA
GCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNAGGAARRWKGGCCAGCAAAGCTACTCTGGAAAGGTGAAGGGGCAGTAGTC
ATACAAGACAATAGTGAAATAAAGGTAGTACCAAGAAGAAAAGCAAAGATCATTAGGGAT
TATGGAAAACAGATGGCAGGTGATGATTGTGTGGCAGGTAGACAGAATGAGGATTAACac
atggaaaagtttagtaaaataccatatgcatgtttcaaagaaagccaaaagatggtttta
tagacctcactttgaaagcatgcatccaagagtaagttcagaagtacacatcccactaga
ggaagctaaattagtaataacaacatattggggtctgcatacaggaNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNaggtgtgaatatcaagcaggacataN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
N
```
For generating this consensus sequence, obtained reads were mapped to HXB2 reference, variants were called (AF >= 0.15) and consensus were generated with bcftools. We used "N" for regions without coverage (this is an aplicon-based sequencing). IUPAC codes were used for with 2 or more haplotypes.

I'm using sierra-web-service for output comparison. The analysis output of sierra-web-service is attached.
[analysis-reports-consensus.zip](https://github.com/PoonLab/sierra-local/files/12147232/analysis-reports-consensus.zip)
 
The siera-local, running on postalign, output json file is also attached .
[consensus_resistance.json.zip](https://github.com/PoonLab/sierra-local/files/12147275/consensus_resistance.json.zip)

**Comment by ArtPoon** [comment]
@WilliamZekaiWang can you update us about this for next week?

**Comment by WilliamZekaiWang** [comment]
I've looked deeper into this issue and I think the problem could be partially on our end. Post-align isn't the issue, it provides the correct data.

The issue seems to lie in the XML files for HIVDB, in our case `HIVDB_9.4.xml`, that we use to get comments about specific drug-resistant mutations. Updating this file with the updater script didn't fix the issue.

Of the incorrectly reported drug resistance mutations, they are all listed as an amino acid mutated to `X` from the nucleotides being `NNN`. I'm pretty sure we match the new amino acid and position to see if we get a hit in the `HIVDB_9.4.xml`. However, since it's ambiguous, sierra-web might ignore these cases. I will try to make Sierra-local ignore reporting potential drug-resistant mutations if `X` includes the consensus amino acid to see if this issue is resolved.

another error I found while looking around is with how sierra-web displays the specific mutation in drug-resistant mutations. Below are the collated drug-resistant mutations for a specific gene.


```
        "PR": [
            "V82VAIMT",
```
compared to sierra-local
```
        "PR": [
            "V82A",
            "V82T"
```
The actual mutation is `V82X` with X being one of `AIMTV`. We do correctly report this in sierra-local. However, the `HIVDB_9.4.xml` file, only holds information on `V82A` and `V82T` drug resistance. Hence, we only state  `V82A` and `V82T` rather than `VAIMT`, under `partialScores` for `drugScores`, conflicting with sierra-web's output.

**Comment by ArtPoon** [comment]
@WilliamZekaiWang can you please generate a few more test cases by modifying a published sequence and running it through HIVdb's web service to capture the JSONs as expected outputs?  Then we can close this out.

**Comment by WilliamZekaiWang** [comment]
There are some issues with the script, but it dealt with the user's sequence fine and reported the drug-resistant scores correctly.

However, there are incorrect drug scores in one of the mutated sequences, where I just took the sample RT file and replaced parts of the sequence with N. As of right now, I'm not too sure what is causing this and I'm still looking into it

**Comment by ArtPoon** [comment]
@WilliamZekaiWang can you please summarize the divergence and labels for test sequences (these are artificial HIV sequences) in this issue?

**Comment by WilliamZekaiWang** [comment]
I have uploaded the test file,  `tests\ambiguous_seq_results.json`, which has N inserted at various positions on a random RT sequence (taken from the sequence in the RT.fa file that is in the readme).

Briefly:
- `RT1_frontN`  has the first 661 Nucleotides replaced with N
- `RT1_endN` has the last 192 Nucleotides replaced with N
- `RT1_middleN` has 354 Nucleotides in the middle replaced with N
- the rest of the sequences are non-mutated RT sequences to check if my script broke any of the original functions

Comparing the scores of the drugs to sierrapy, `RT_endN` has an incorrect drug resistance report at `H221Y`, where sierrapy had nothing.

`RT_endN` also has incorrect drug scores for the following drugs
```
            "DOR": 10.0,
            "EFV": 10.0,
            "ETR": 10.0,
            "NVP": 15.0,
            "RPV": 15.0
```
where sierrapy reported 0 for all these drugs

**Comment by ArtPoon** [comment]
We're going to have to dig into the sierrapy code to check how they are handling positions without coverage (missing data).  I had thought that they would take the conservative approach (assume the worst case, i.e., resistance mutations are present).

**Comment by WilliamZekaiWang** [comment]
I didn't have much luck digging up the code from standford's repository

I made scripts to find the differences and errors and pushed them to branch iss72. I also think I found the error. 

The above extra mutations were all caused by a singular mutation that we assign as `H221NHDY` but sierrapy assigns as `H221X`. It seems like it's safe to assume that any ambiguous mutations labeled with `X` are ignored in their drug score calculation.

Unfortunately, this likely means that either, there's post alignment after post-align, or we are using a different version than sierrapy.
